pysensors/utils/_constraints.py

```diff
@@ -64,47 +66,350 @@ def get_constrained_sensors_indices_linear(x_min, x_max, y_min, y_max,df):
+        n_samples, n_features = self.data.shape
+        n_sensors = len(sensors)
+        if isinstance(self.data,np.ndarray):
+            xTop = np.mod(sensors,np.sqrt(n_features))
+            yTop = np.floor(sensors/np.sqrt(n_features))
+            data = np.vstack([sensors,xTop,yTop]).T
+            for ind,i in enumerate(range(len(xTop))):
+                plt.annotate(f"{str(ind)}",(xTop[i],yTop[i]),xycoords='data',
+                    xytext=(-20,20), textcoords='offset points',color="r",fontsize=12,
+                    arrowprops=dict(arrowstyle="->", color='black'))
+        elif isinstance(self.data,pd.DataFrame):
+            X,Y = self.data[self.X_axis], self.data[self.Y_axis]
+            xTop, yTop = get_coordinates_from_indices(sensors,self.data)
+            data = np.vstack([sensors,xTop,yTop]).T
+            for _,i in enumerate(range(len(xTop))):
+                plt.annotate(f"{str(i)}",(xTop.to_numpy()[i]*100,yTop.to_numpy()[i]*100),xycoords='data',
+                    xytext=(-20,20), textcoords='offset points',color="r",fontsize=12,
+                    arrowprops=dict(arrowstyle="->", color='black'))
+    
+class Circle(BaseConstraint):
+    '''
+    General class for dealing with circular user defined constraints.
+    Plotting, computing constraints functionalities included. 
+    '''
+    def __init__(self,center_x,center_y,radius, **kwgs):
+        super().__init__(**kwgs)
+        '''
+        Attributes
+        ----------
+        center_x : float,
+            x-coordinate of the center of circle
+        center_y : float,
+            y-coordinate of the center of circle
+        radius : float,
+            radius of the circle
+        '''
+        self.center_x = center_x
+        self.center_y = center_y
+        self.radius = radius
+        
+    def draw(self,ax):
+        c = patches.Circle((self.center_x, self.center_y), self.radius, fill = False, color = 'r', lw = 2)
+        ax.add_patch(c)
+        ax.autoscale_view()
+        
+        
+    def constraint_function(self,x_all_unc, y_all_unc):
+        '''
+        To be Filled
+        '''
+        return ((x_all_unc-self.center_x)**2 + (y_all_unc-self.center_y)**2) - self.radius**2
+        
+           
+class Line(BaseConstraint):
+    '''
+    To be filled
+    '''
+    def __init__(self,x1,x2,y1,y2,**kwgs):
+        super().__init__(**kwgs)
+        '''
+        Attributes
+        ----------
+        x1 : float,
+            x-coordinate of one end-point of the line
+        x2 : float,
+            x-coordinate of the other end-point of the line
+        y1 : float,
+            y-coordinate of one end-point of the line
+        y2 : float,
+            y-coordinate of the other end-point of the line
+        '''
+        self.x1 = x1
+        self.x2 = x2
+        self.y1 = y1
+        self.y2 = y2
+        
+    def draw(self,ax):
+        '''
+        To be Filled
+        '''
+        ax.plot([self.x1,self.x2],[self.y1,self.y2],'-r')
```

Comment:
Hm, I don't think this gives an indication which side is the constrained region.

Reply:
Yes, I still do not know exactly how do we want to handle on which side of the line is that constraint. Maybe we can have a discussion on this with @Jimmy-INL .